doc/specs/#4999 - Improved keyboard handling in Conpty.md

```diff
@@ -0,0 +1,452 @@
+* Wouldn't really expect client apps to ever _need_ this format, but it could
+  always be possible for them to need it in the future.
+
+#### Pros:
+* Definitely gets us all the information that we need.
+* Can handle solo modifiers
+* Can handle keydown and keyup separately
+* We can make the sequence however we want to parse it.
+
+#### Cons:
+* No reference implementation, so we'd be flying blind
+* We'd be defining our own VT sequences for these, which we've never done
+  before. This was _inevitable_, however, this is still the first time we'd be
+  doing this.
+* Something about Microsoft extending well-defined protocols being bad 😆
+* By having the Terminal send all input as _this protocol_, VT Input passthrough
+  to apps that want VT input won't work anymore for the Terminal. That's _okay_
+
+### kitty extension
+[Reference](https://sw.kovidgoyal.net/kitty/protocol-extensions.html#keyboard-handling)
+
+#### Pros:
+* Not terribly difficult to decode
+* Unique from anything else we'd be processing, as it's an APC sequence
+  (`\x1b_`)
+* From their docs: > All printable key presses without modifier keys are sent
+  just as in the normal mode. ... For non printable keys and key combinations
+  including one or more modifiers, an escape sequence encoding the key event is
+  sent
+   - I think like this. ASCII and other keyboard layout chars (things that would
+     hit `SendChar`) would still just come through as the normal char.
+
+#### Cons:
+* Their encoding table is _mad_. [Look at
+  this](https://sw.kovidgoyal.net/kitty/key-encoding.html). What order is that
+  in? Obviously the first column is sorted alphabetically, but the mapping of
+  key->char is in a seemingly bonkers order.
+* I can't get it working locally, so hard to test 😐
+* They do declare the `fullkbd` terminfo capability to identify that they
+  support this mode, but I'm not sure anyone else uses it.
+  - I'm also not sure that any _client_ apps are reading this currently.
+* This isn't designed to be full `KEY_EVENT`s - where would we put the scancode
+  (for apps that think that's important)?
+  - We'd have  to extend this protocol _anyways_
+
+### `xterm` "Set key modifier options"
+Notably looking at
+[`modifyOtherKeys`](https://invisible-island.net/xterm/manpage/xterm.html#VT100-Widget-Resources:modifyOtherKeys).
+
+#### Pros:
+* `xterm` implements this so there's a reference implementation
+* relatively easy to parse these sequences. `CSI 27 ; <modifiers> ; <key> ~`
+
+#### Cons:
+* Only sends the sequence on key-up
+* Doesn't send modifiers all on their own
+
+### `DECPCTERM`
+[VT100.net doc](https://vt100.net/docs/vt510-rm/DECPCTERM.html)
+
+#### Pros:
+* Enables us to send key-down and key-up keys independently
+* Enables us to send modifiers on their own
+* Part of the VT 510 standard
+
+#### Cons:
+* neither `xterm` nor `gnome-terminal` (VTE) seem to implement this. I'm not
+  sure if anyone's got a reference implementation for us to work with.

```

Comment:
Not that it matters at this point, but you might be interested to know that Kermit claims to support a PCTERM mode, although it sounds like it needs to be turned on manually (the relevant docs are [here](http://www.columbia.edu/kermit/k95manual/#c7pcterm)). I haven't had a chance to try it out myself, so I don't know how well it works.